### PR TITLE
Confluent Cloud Integration

### PR DESCRIPTION
**Pull request by dmccoystephenson**
## Purpose
These changes allow the ACM to communicate with Confluent Cloud.

## Enabling Confluent Cloud Integration
* Set the KAFKA_TYPE, CONFLUENT_KEY and CONFLUENT_SECRET environment variables and pass them to the submodules of the ODE. This can be done by using the docker-compose-confluent-cloud file in place of the usual docker-compose file.
* Alter the DOCKER_HOST_IP environment variable.

### CODE REVIEW

**Comment by dmccoystephenson** [comment]
These changes have been tested to be backwards compatible. When the KAFKA_TYPE environment variable is not set or is set to something other than "CONFLUENT", the ACM will attempt to work with a local kafka instance as usual.

File: Dockerfile [hunk 1]
```diff
@@ -13,13 +13,16 @@ RUN apt-get update && apt-get install -y software-properties-common wget git mak
...
-RUN wget https://cmake.org/files/v3.7/cmake-3.7.2.tar.gz && tar -xvf cmake-3.7.2.tar.gz
-RUN cd cmake-3.7.2 && ./bootstrap && make && make install && cd /home
+RUN apt-get install -y cmake
 
 # install libtool and automake
-RUN apt-get update && apt-get install -y automake libtool
+RUN apt-get install -y automake libtool
 
 # Install librdkafka.
+RUN apt-get install -y libsasl2-dev
```
Comment (payneBrandon) [inline Dockerfile:22]: Let's combine these into a single "RUN" command to make this a bit cleaner

File: Dockerfile [hunk 2]
```diff
@@ -13,13 +13,16 @@ RUN apt-get update && apt-get install -y software-properties-common wget git mak
 #RUN apt-get update && apt-get install -y nano
```
Comment (payneBrandon) [inline Dockerfile:13]: If we're not using the vim/nano tools, let's remove from the Dockerfile while we're in here

File: docker-compose-confluent-cloud.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+version: '2'
```
Comment (payneBrandon) [inline docker-compose-confluent-cloud.yml:1]: Let's format this in the same way as the other docker-compose.yml file, which spins up the standalone version along with a Kafka image. Then when we create a PR to the main USDOT repo we can include documentation on spinning up multiple containers targeting Confluent. 

Currently, these services are set to depend on "kafka" image which doesn't exist within this yml file.

File: src/acm.cpp [hunk 4]
```diff
@@ -304,20 +313,24 @@ bool ASN1_Codec::topic_available( const std::string& topic ) {
...
-        if (!r) ilogger->warn( "Metadata did not contain topic: {}.", topic );
+        if (!r) {
+            ilogger->warn( "Metadata did not contain topic: {}.", topic );
+        }
 
     } else {
         elogger->error( "cannot retrieve consumer metadata with error: {}.", err2str(err) );
     }
     
+    // std::cout << "The topic_available() method has finished executing. Return value: " << r << std::endl;
```
Comment (payneBrandon) [inline src/acm.cpp:329]: Let's clean out this commented statement

File: src/acm.cpp [hunk 5]
```diff
@@ -72,6 +72,15 @@
 #include <unistd.h>
 #endif
 
+const char* getEnvironmentVariable(const char* variableName) {
+    const char* toReturn = std::getenv(variableName);
+    if (!toReturn) {
+        std::cout << "Something went wrong attempting to retrieve the environment variable " << variableName << std::endl;
```
Comment (payneBrandon) [inline src/acm.cpp:78]: Let's throw this to the logger rather than cout, error level.

File: src/acm.cpp [hunk 6]
```diff
@@ -474,6 +484,26 @@ bool ASN1_Codec::configure() {
...
+        std::string username = getEnvironmentVariable("CONFLUENT_KEY");
+        std::string password = getEnvironmentVariable("CONFLUENT_SECRET");
+
+        // set up config
+        conf->set("bootstrap.servers", getEnvironmentVariable("DOCKER_HOST_IP"), error_string);
+        conf->set("security.protocol", "SASL_SSL", error_string);
+        conf->set("sasl.mechanism", "PLAIN", error_string);
+        conf->set("sasl.username", username.c_str(), error_string);
+        conf->set("sasl.password", password.c_str(), error_string);
+        // conf->set("debug", "all", error_string);
```
Comment (payneBrandon) [inline src/acm.cpp:500]: Let's remove.

File: src/acm.cpp [hunk 7]
```diff
@@ -546,11 +576,11 @@ bool ASN1_Codec::configure() {
     }
 
     ilogger->trace("{}: finished.", fnname );
+    // print_configuration();
```
Comment (payneBrandon) [inline src/acm.cpp:579]: Let's remove

**Review by payneBrandon — CHANGES_REQUESTED** [review]
Looks good, just a few comments. Along with the migration to an installed librdkafka instance, we should remove the librdkafa submodule so there's no confusion.

File: src/acm.cpp (hunk 5 above)
Comment (dmccoystephenson) [reply src/acm.cpp]: This method has been moved to the ASN1_Codec class so the internal logger could be utilized.

File: Dockerfile (hunk 1 above)
Comment (dmccoystephenson) [reply Dockerfile]: These commands have been combined.

File: Dockerfile (hunk 2 above)
Comment (dmccoystephenson) [reply Dockerfile]: These lines have been removed.

File: src/acm.cpp (hunk 4 above)
Comment (dmccoystephenson) [reply src/acm.cpp]: Removed.

File: src/acm.cpp (hunk 6 above)
Comment (dmccoystephenson) [reply src/acm.cpp]: Removed.

File: src/acm.cpp (hunk 7 above)
Comment (dmccoystephenson) [reply src/acm.cpp]: Removed.

**Review by drewjj — APPROVED** [review]
Things look good, we will need to write some documentation on this. I plan to include the documentation in the ODE's wiki.  I will meet with you to go over that today.

File: docker-compose-confluent-cloud.yml (hunk 3 above)
Comment (payneBrandon) [reply docker-compose-confluent-cloud.yml:1]: When we make the PR for the ODE main module, we'll include changed to that docker-compose file that will include these separated decode/encode with optional CC integration

**Comment by dmccoystephenson** [comment]
A section on CC integration has been added to the README.

**Comment by dmccoystephenson** [comment]
An image has been pushed to our cluster with these updates.

**Review by payneBrandon — APPROVED** [review]
Looks good to me. I added a `build_local.sh` script to allow running all the build commands locally easier. If you think that looks good, we can go ahead and merge.

**Comment by dmccoystephenson** [comment]
Looks good to me!

**Pull request merged**